[GStreamer][MSE] Support MP3 with ID3 tags

eocanha opened the pull request:
#### a4e8e5b80d172b1bd4f7afa2b3099d9e1bc81186
<pre>
[GStreamer][MSE] Support MP3 with ID3 tags
<a href="https://bugs.webkit.org/show_bug.cgi?id=291414">https://bugs.webkit.org/show_bug.cgi?id=291414</a>

Reviewed by Alicia Boya Garcia.

If an MP3 with ID3 tags is appended to a SourceBuffer, AppendPipeline&apos;s
typefind detects it as application/x-id3 instead of audio/mpeg and can&apos;t
process it.

This commit takes care of that case by (optionally) creating an id3demux
that takes care of the ID3 tags (just ignores them) and produces
audio/mpeg at its output, so the processing can continue normally.

We can&apos;t know if id3demux can be created or not at AppendPipeline
instantiation time. In case the SourceBuffer mimetype is audio/mpeg, we
need to listen to the have-type signal from typefind and defer the
creation, configuration and connection of the demuxer (or identity, if
id3demux is not needed) until the signal happens.

The code to configure the demuxer has been refactored into its own
method, so it can be reused from the regular and the deferred demuxer
configuration codepaths.

* LayoutTests/media/content/silence_with_id3_tags.mp3: Added.
* LayoutTests/media/media-source/media-source-play-mp3-id3-tag-expected.txt: Added.
* LayoutTests/media/media-source/media-source-play-mp3-id3-tag.html: Added.
* Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
(WebCore::AppendPipeline::configureOptionalDemuxerFromAnyThread): Refactored code to configure the demuxer. The configuration code has been enhanced to detect the case of demuxers with static pads (such as id3demux) and treat them in the same immediate configuration way as identity.
(WebCore::AppendPipeline::AppendPipeline): If the SourceBuffer mimetype is audio/mpeg, listen to have-type and create and configure the demuxer/identity there (just the first time the signal is triggered).
* Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.h: Added new configureOptionalDemuxerFromAnyThread() private method.

Canonical link: <a href="https://commits.webkit.org/294106@main">https://commits.webkit.org/294106@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c2344060e65149c898b2b58a1955dc6a073b4186

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100862 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20514 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10813 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105999 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51451 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/102903 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20823 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28988 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/76805 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33841 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103869 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/16011 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91104 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57157 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15822 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/9111 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50828 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/85722 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9187 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108354 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27980 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/20550 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/85767 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28343 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87306 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85311 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21709 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/30023 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7753 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/22001 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27915 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/33173 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/27726 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/31044 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29284 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

philn commented on Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
```diff
@@ -154,8 +187,45 @@ AppendPipeline::AppendPipeline(SourceBufferPrivateGStreamer& sourceBufferPrivate
+            if (capsString.startsWith("application/x-id3"_s))
+                appendPipeline->m_demux = makeGStreamerElement("id3demux"_s);
+            else if (capsString.startsWith("audio/mpeg"_s))
+                appendPipeline->m_demux = makeGStreamerElement("identity"_s);
```
There is no need for a conversion to string. Get the first structure from the caps and use `gst_structure_has_name()`

eocanha replied:
Done, but still, when I later need to print the name by using `gst_structure_get_name()`, webkit-style-check nags me about using `StringView gstStructureGetName()`. This forces me to use `gstStructureGetName(capsStructure).toStringWithoutCopying().utf8().data()`. I don't know how wasteful all those conversions are. I always have the impression that C++ is more about churning type conversions than about doing useful computations. :-/

philn commented on Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
```diff
@@ -154,8 +187,45 @@ AppendPipeline::AppendPipeline(SourceBufferPrivateGStreamer& sourceBufferPrivate
+
+        // Add_many will take ownership of a reference. That's why we used an assignment before.
+        gst_bin_add_many(GST_BIN(m_pipeline.get()), m_appsrc.get(), m_typefind.get(), nullptr);
+        gst_element_link_many(m_appsrc.get(), m_typefind.get(), nullptr);
```
gst_element_link()

csaavedra commented on Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
```diff
@@ -104,6 +104,39 @@ static void assertedElementSetState(GstElement* element, GstState desiredState)
+    // We try to detect special cases of demuxers that have a single static src pad, such as id3demux.
+    GRefPtr<GstPad> demuxerSrcPad = adoptGRef(gst_element_get_static_pad(m_demux.get(), "src"));
+    auto classifiers = elementClass.split('/');
+    if (classifiers.contains("Demuxer"_s) && !demuxerSrcPad) {
```
From what I see if `demuxerSrcPad` is not `nullptr` there is no need to create a string and split it. This could probably be improved.

philn replied:
It can be null, not all demuxers output a single pad named "src".

csaavedra replied:
Yes, but when it's not, you are creating a string and splitting it for no use. I'd refactor this code.

eocanha replied:
Explaining aloud to myself my understanding of the comments, so I don't forget when I come back at this in one week:
- Evaluate `!demuxerSrcPad` as first element in short-circuit (or alone) in the if
- Move elementClass and classifiers initialization inside the branch, or even inside the condition
- Make sure that the `else` code runs in the `!!demuxerSrcPad` case and also in case the element is not a "Demuxer".

Checking only `if (demuxerSrcPad)` and swapping the then and else branches could do the trick. If the element doesn't have a "src" pad (new else branch) and is not a demuxer, then we can just assert or something, I guess, because that would be an unsupported case (for the limited set of formats that AppendPipeline can support: webm, mp4, raw mpeg or aac stream).

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
```diff
@@ -154,8 +187,45 @@ AppendPipeline::AppendPipeline(SourceBufferPrivateGStreamer& sourceBufferPrivate
+            ASSERT(!isMainThread());
+
+            // We don't want to create the demuxer twice if the type changes for whatever reason,
+            // and this is cheaper than storing the handler id and disconnecting the handler.
```
I would remove the second line. The first line is enough justification, whereas the second line may give a future reader the impression that storing a handler ids is expensive, which it is not.

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
```diff
@@ -154,8 +187,45 @@ AppendPipeline::AppendPipeline(SourceBufferPrivateGStreamer& sourceBufferPrivate
+                return;
+
+            auto capsString = String::fromUTF8(GUniquePtr<char>(gst_caps_to_string(caps)).get());
+            GST_DEBUG_OBJECT(appendPipeline->pipeline(), "Creating demuxer for %s caps", capsString.utf8().data());
```
This log would be much more useful if it told you what demuxer it will create.

You can accomplish that like this (pseudo-code):

```
ASCIILiteral demuxerName = NULL;
if is_id3 {
  demuxerName = "id3demux";
} else if is_mpeg {
  demuxerName = "identity"
}
if !demuxerName {
  GST_ELEMENT_ERROR(pipeline, "Found caps incompatible with %s: %s", mimeType, caps);
  return;
}
GST_DEBUG_OBJECT(pipeline, "Creating demuxer %s for %s caps", demuxerName, caps);
m_demux = makeGStreamerElement(demuxerName);
```

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
```diff
@@ -154,8 +187,45 @@ AppendPipeline::AppendPipeline(SourceBufferPrivateGStreamer& sourceBufferPrivate
+            else
+                GST_ERROR_OBJECT(appendPipeline->pipeline(), "Unsupported caps for audio/mpeg mimetype: %s", capsString.utf8().data());
+
+            ASSERT(m_demux);
```
Erroring in the case of invalid caps for audio/mpeg SourceBuffer type must not be done via an assertion failure, as it is an error an user can trigger with JS alone (e.g. by creating a SourceBuffer with audio/mpeg type and appending MP4 instead).

Instead, when that error happens, we should make the AppendPipeline and SourceBuffer error out to the user. You can have a look at `handleErrorConditionFromStreamingThread()` for an example of what that looks like. However, you don't need to do all of that or call that function, instead you can use `GST_ELEMENT_ERROR()` with the pipeline as element.

`GST_ELEMENT_ERROR()`  will post an error message to the bus, which will be caught synchronously `handleErrorSyncMessage()` which will log it as warning (as it's not necessarily WebKit's fault), call `handleErrorConditionFromStreamingThread()` and dump the pipeline.

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
```diff
@@ -154,8 +187,45 @@ AppendPipeline::AppendPipeline(SourceBufferPrivateGStreamer& sourceBufferPrivate
+
+            appendPipeline->configureOptionalDemuxer();
+
+            // Add_many will take ownership of a reference. That's why we used an assignment before.
```
"take ownership of a reference" is a bit **ambiguous**: It could be read (in this case correctly) to mean that the function will increase the refcount of the object, but it may also be read (in this case incorrectly) to mean that the ownership is being moved.

For context, this is what really happens:

`gst_bin_add_many()` and `gst_bin_add()` take transfer floating arguments. Transfer floating in argument means:

* if the object is floating, it works as transfer full but the floating flag is removed. Caller code no longer owns the reference.
* if the object is not floating, it works as transfer none and the function being called acquires a new owning reference. Caller code still owns a reference.

Elements are born floating. When you pass it immediately after to the constructor of GRefPtr (or equivalently, assign with `=` to a GRefPtr), **the floating flag is removed** and GRefPtr becomes the only owner of the object. Additionally GRefPtr's are forbidden from holding floating references.

Hence, when passing GRefPtr to transfer floating arguments (e.g. `gst_bin_add()`) `.get()` must be used and it will work with equivalent transfer none semantics.

---

Now, context aside: I think **the comment as it stands risks more confusion than helping**, as the code below is not particularly surprising on its own: What would you inadvertedly do other than `.get()` if you were not aware of these semantics?

* Surely not `GRefPtr.leakPtr()` as that explicitly and safely transfers your reference away.
* Not making any GRefPtr at all and passing the raw pointer from the element constructor? That would work safely too as long as you don't unref the pointer or adoptGRef() it after.

Turns out in this particular case GRefPtr is actually giving us enough guardrails. It's nice when things work as expected for once!

---

Later, I just realized this comment is not new, but it is already in the original AppendPipeline commit from 2016! The analysis still stands and I still think it makes sense to remove it.

Hopefully the new [GRefPtr doc comments](https://github.com/WebKit/WebKit/pull/44370) helps understanding floating shenanigans with GRefPtr better than that comment for that particular usage of that particular function.

eocanha replied:
Thanks for the explanation. I still prefer to correct the comment and leave it, as a gift for future developers ("It's dangerous to work alone with refcounting semantics, take this explanation!")

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
```diff
@@ -154,8 +187,45 @@ AppendPipeline::AppendPipeline(SourceBufferPrivateGStreamer& sourceBufferPrivate
+            assertedElementSetState(appendPipeline->m_demux.get(), GST_STATE_PLAYING);
+        }), this);
+
+        // Add_many will take ownership of a reference. That's why we used an assignment before.
```
Same thing as the previous comment.

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
```diff
@@ -104,6 +104,39 @@ static void assertedElementSetState(GstElement* element, GstState desiredState)
     }
 }
 
+void AppendPipeline::configureOptionalDemuxer()
```
`configureOptionalDemuxerFromAnyThread()`

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
```diff
@@ -154,8 +187,45 @@ AppendPipeline::AppendPipeline(SourceBufferPrivateGStreamer& sourceBufferPrivate
+        gst_element_link_many(m_appsrc.get(), m_typefind.get(), nullptr);
+
+        assertedElementSetState(m_pipeline.get(), GST_STATE_PLAYING);
+        return;
```
Returning early is not the right call here; instead we want to keep the code as linear as possible to make it easier to analyze.

~~As an example of a bug consequence of this early return, consider the `demuxer data entering` probe code just below. It runs for all branches except this one, even though it would be just as helpful.~~ Edit: I just realized, this doesn't make so much sense since the `m_demux` is null at that point, so that would have to be delayed as well...

Here is an idea to make it more linear:

1. Explicitly document that `m_demux` can be null if it's not known at this point whether an id3demux will be needed.
2. Remove the code for adding and linking from here, as well as the set state and the return.
3. Scroll down to the original code for adding and linking elements and wrap it in `if (m_demux)`. If true, we add and link appsrc, identity and demux. If false, we only add and link appsrc and identity.
   This may seem a trivial change as we would end up calling the same functions with the same values in the same cases. However, after the change it is much more obvious that those two possible outcomes exist and how they're handled, as you can see both of them together and explicitly cased.

eocanha replied:
We can refactor both branches and add all of m_appsrc, m_typefind and m_demux at that point, even if m_demux is null. That null value would act as the varargs list guard 🙃 (with the proper explaining comment, of course).

ntrrgc reviewed (changes requested):
The logic makes sense, with the minor exception of the handling of unsupported caps as I explained. Otherwise, the changes needed are about code structure.

It makes me a bit sad that we have to break the linearity of the append pipeline a little bit to accomodate id3demux, but the changes explained it should mitigate the added complexity just enough.

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
```diff
@@ -154,47 +189,69 @@ AppendPipeline::AppendPipeline(SourceBufferPrivateGStreamer& sourceBufferPrivate
+
+            appendPipeline->configureOptionalDemuxerFromAnyThread();
+
+            // The added element has a sunk reference after being assigned to the GRefPtr, so the transfer-floating
```
The added element had its floating reference sunk after being assigned to the GRefPtr

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
```diff
@@ -154,47 +189,69 @@ AppendPipeline::AppendPipeline(SourceBufferPrivateGStreamer& sourceBufferPrivate
+            GST_DEBUG_OBJECT(appendPipeline->pipeline(), "Creating %s demuxer for %s caps",
+                demuxerElementName.characters(), gstStructureGetName(capsStructure).toStringWithoutCopying().utf8().data());
+            appendPipeline->m_demux = makeGStreamerElement(demuxerElementName);
+
```
Nit: I would remove this empty line so that setting the new demuxer and asserting the demuxer is not null are part of the same block.

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
```diff
@@ -154,47 +189,69 @@ AppendPipeline::AppendPipeline(SourceBufferPrivateGStreamer& sourceBufferPrivate
+                return;
+            }
+
+            GST_DEBUG_OBJECT(appendPipeline->pipeline(), "Creating %s demuxer for %s caps",
```
Nit: Since caps strings can be very long, I generally prefer leaving them at the end where possible, e.g:  `Creating %s demuxer for caps: %s` -- it helps with copying and pasting.

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
```diff
@@ -154,47 +189,69 @@ AppendPipeline::AppendPipeline(SourceBufferPrivateGStreamer& sourceBufferPrivate
+        return;
+    }
+
+    // m_demux might be null at this point if there's a typefind pending to identify the proper demuxer to be used
+    // (see the audio/mpeg case right above).
```
Thanks for this comment.

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
```diff
@@ -154,47 +189,69 @@ AppendPipeline::AppendPipeline(SourceBufferPrivateGStreamer& sourceBufferPrivate
-    // Add_many will take ownership of a reference. That's why we used an assignment before.
+    // The added elements have a sunk reference after being assigned to the GRefPtr, so the transfer-floating
+    // parameters are working as transfer-none here.
+    // Note that m_demux may be null at this point, so the variable argument list would ignore it (m_demux would
```
Sneaky smart. I've checked https://en.cppreference.com/w/c/language/variadic and I see no requirement va_arg() being called for *all* the arguments, so it should be safe.

ntrrgc reviewed (approved):
LGTM.

eocanha committed:
[GStreamer][MSE] Support MP3 with ID3 tags
https://bugs.webkit.org/show_bug.cgi?id=291414

Reviewed by Alicia Boya Garcia.

If an MP3 with ID3 tags is appended to a SourceBuffer, AppendPipeline's
typefind detects it as application/x-id3 instead of audio/mpeg and can't
process it.

This commit takes care of that case by (optionally) creating an id3demux
that takes care of the ID3 tags (just ignores them) and produces
audio/mpeg at its output, so the processing can continue normally.

We can't know if id3demux can be created or not at AppendPipeline
instantiation time. In case the SourceBuffer mimetype is audio/mpeg, we
need to listen to the have-type signal from typefind and defer the
creation, configuration and connection of the demuxer (or identity, if
id3demux is not needed) until the signal happens.

The code to configure the demuxer has been refactored into its own
method, so it can be reused from the regular and the deferred demuxer
configuration codepaths.

* LayoutTests/media/content/silence_with_id3_tags.mp3: Added.
* LayoutTests/media/media-source/media-source-play-mp3-id3-tag-expected.txt: Added.
* LayoutTests/media/media-source/media-source-play-mp3-id3-tag.html: Added.
* Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
(WebCore::AppendPipeline::configureOptionalDemuxerFromAnyThread): Refactored code to configure the demuxer. The configuration code has been enhanced to detect the case of demuxers with static pads (such as id3demux) and treat them in the same immediate configuration way as identity.
(WebCore::AppendPipeline::AppendPipeline): If the SourceBuffer mimetype is audio/mpeg, listen to have-type and create and configure the demuxer/identity there (just the first time the signal is triggered).
* Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.h: Added new configureOptionalDemuxerFromAnyThread() private method.

Canonical link: https://commits.webkit.org/294106@main
